### PR TITLE
Implement a completion for View parameter of format cmdlets

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Now works `dir | ft -View <Tab>`.
It has always been a problem to find out which format views exist.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7462
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 1]
```diff
@@ -2284,6 +2285,14 @@ internal static string ConcatenateStringPathArguments(CommandElementAst stringAs
                 case "Measure-Object":
                 case "Sort-Object":
                 case "Where-Object":
+                    {
+                        if (parameterName.Equals("Property", StringComparison.OrdinalIgnoreCase))
+                        {
+                            NativeCompletionMemberName(context, result, commandAst);
+                        }
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:2292]: Is this an unrelated change?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 2]
```diff
@@ -3787,44 +3800,78 @@ private static void NativeCompletionPathArgument(CompletionContext context, stri
...
+            Dictionary<string, AstParameterArgumentPair> boundArguments,
+            List<CompletionResult> result,
+            CommandAst commandAst,
+            string commandName)
+        {
+            IEnumerable<PSTypeName> prevType = NativeCommandArgumentCompletion_InferTypesOfArgument(boundArguments, commandAst, context, "InputObject");
+
+            if (prevType is not null)
+            {
+                var inferTypeNames = prevType.Select(t => t.Name).ToArray();
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:3871]:
```suggestion
                string[] inferTypeNames = prevType.Select(t => t.Name).ToArray();
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:2292]: No, we have to split the code path.

**Review by rjmholt — COMMENTED** [review]
This change looks good, just have some style requests

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 3]
```diff
@@ -3787,44 +3800,78 @@ private static void NativeCompletionPathArgument(CompletionContext context, stri
             result.Add(CompletionResult.Null);
         }
 
-        private static void NativeCompletionMemberName(CompletionContext context, List<CompletionResult> result, CommandAst commandAst)
+        private static IEnumerable<PSTypeName> GetInferenceTypes(CompletionContext context, List<CompletionResult> result, CommandAst commandAst)
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:3803]: `result` is in parameter list.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 4]
```diff
@@ -3787,44 +3800,78 @@ private static void NativeCompletionPathArgument(CompletionContext context, stri
...
+
+        private static void NativeCompletionMemberName(CompletionContext context, List<CompletionResult> result, CommandAst commandAst)
+        {
+            IEnumerable<PSTypeName> prevType = GetInferenceTypes(context, result, commandAst);
+            if (prevType is not null)
+            {
+                CompleteMemberByInferredType(context.TypeInferenceContext, prevType, result, context.WordToComplete + "*", filter: IsPropertyMember, isStatic: false);
+            }
+
+            result.Add(CompletionResult.Null);
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:3857]: It is used as a terminator.  It is added in all code paths.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 5]
```diff
@@ -5244,6 +5291,46 @@ private static bool IsInDscContext(ExpressionAst expression)
...
+                "Format-Table" => typeof(TableControlBody),
+                "Format-List" => typeof(ListControlBody),
+                "Format-Wide" => typeof(WideControlBody),
+                "Format-Custom" => typeof(ComplexControlBody),
+                _ => null
+            };
+
+            Diagnostics.Assert(controlBodyType is null, "This should never happen unless a new Format-* cmdlet is added");
+
+            HashSet<string> uniqueNames = new();
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5314]: Done.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 6]
```diff
@@ -5244,6 +5291,46 @@ private static bool IsInDscContext(ExpressionAst expression)
             return Ast.GetAncestorAst<ConfigurationDefinitionAst>(expression) != null;
         }
 
+        private static void CompleteFormatViewByInferredType(TypeInferenceContext context, string[] inferredTypeNames, List<CompletionResult> results, string commandName)
+        {
+            var db = context.ExecutionContext.FormatDBManager.GetTypeInfoDataBase();
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5296]: Done.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 7]
```diff
@@ -5244,6 +5291,46 @@ private static bool IsInDscContext(ExpressionAst expression)
...
+                "Format-List" => typeof(ListControlBody),
+                "Format-Wide" => typeof(WideControlBody),
+                "Format-Custom" => typeof(ComplexControlBody),
+                _ => null
+            };
+
+            Diagnostics.Assert(controlBodyType is null, "This should never happen unless a new Format-* cmdlet is added");
+
+            HashSet<string> uniqueNames = new();
+            foreach (ViewDefinition vd in db.viewDefinitionsSection.viewDefinitionList)
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5315]: Done.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 8]
```diff
@@ -5244,6 +5291,46 @@ private static bool IsInDscContext(ExpressionAst expression)
...
+            {
+                if (vd is not null && controlBodyType == vd.mainControl.GetType() && vd.appliesTo is not null)
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5317]: This looks too tricky for me :-)

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 9]
```diff
@@ -5244,6 +5291,46 @@ private static bool IsInDscContext(ExpressionAst expression)
...
+                {
+                    foreach (var applyTo in vd.appliesTo.referenceList)
+                    {
+                        foreach (var inferredTypeName in inferredTypeNames)
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5321]: Removed there `var`.

**Comment by iSazonov** [comment]
Rebased to move to latest .Net.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:3801]: > result is in parameter list.

We pass a parameter called `result` in, which in other methods is used with a `void` method to add results to instead. But in this method we now return values, so what purpose does the `result` argument serve?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Well you can get a performance increase and a readability improvement by checking the `null` property first:

```csharp
if (viewDefinition is not null && viewDefinition.appliesTo is not null ...)
{
}
```

But then it's a pretty natural step to condense those two checks down into a null-coalesced check.

The way it's written currently `vd` is not self-describing and `vd.mainControl.GetType()` is doing reflection that might not be needed if `vd.appliedTo` is `null`.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:3801]: Ah, you ask to remove the parameter :-) Yes, now we can remove it from the private method.

Done.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Thanks for clarify! I did not think many about performance in the functional PR but I added your proposal although personally I don't like null-conditional operators especially in conditions.

**Review by rjmholt — APPROVED** [review]

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 10]
```diff
@@ -301,6 +301,37 @@ Describe "TabCompletion" -Tags CI {
...
+        ) {
+            param($cmd, $expected)
+
+            # The completion is based on OutputTypeAttribute() of the cmdlet.
+            $res = TabExpansion2 -inputScript "Get-ChildItem | $cmd -View " -cursorColumn "Get-ChildItem | $cmd -View ".Length
+            $completionText = $res.CompletionMatches.CompletionText | Sort-Object
+            $completionText -join ' ' | Should -BeExactly $expected
+        }
+
+        It 'Should complete $processList = Get-Process; $processList | <cmd> -View' -TestCases (
```
Comment (SteveL-MSFT) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:319]: I don't think there's a restriction that a view name can't have whitespace.  Can we have a synthetic test with a view name that has whitespace?  Do the quotes get added?

Comment (iSazonov) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:376]: Good catch!
I added quoting if special characters are in view name.

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 11]
```diff
@@ -301,6 +301,94 @@ Describe "TabCompletion" -Tags CI {
...
+            @{ cmd = 'Format-Table'; expected = "'R A M' Priority process ProcessModule ProcessWithUserName StartTime" },
```
Comment (SteveL-MSFT) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:377]: I recall when I did the completer for enums I had to handle both single and double quote depending on what the user may have started typing.  Can you add a variation of the test for both:

- ... | format-list -view "R<tab>
- ... | format-list -view 'R<tab>

Comment (iSazonov) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1]: I was sure it was a generic. :-(

Fixed.

**Comment by iSazonov** [comment]
I wonder why WindowsCompatibility catch processViewDefinition.ps1xml from the new test. 
/cc @anmenaga 

Also I wonder why the processViewDefinition.ps1xml is not deleted automatically and manually :-(

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 12]
```diff
@@ -5776,7 +5776,11 @@ private static void CompleteFormatViewByInferredType(TypeInferenceContext contex
                             // We use 'StartsWith()' because 'applyTo.Name' can look like "System.Diagnostics.Process#IncludeUserName".
                             if (applyTo.name.StartsWith(inferredTypeName, StringComparison.OrdinalIgnoreCase) && uniqueNames.Add(viewDefinition.name))
                             {
-                                results.Add(new CompletionResult(viewDefinition.name, viewDefinition.name, CompletionResultType.Text, viewDefinition.name));
+                                string completionText = viewDefinition.name.IndexOfAny(s_charactersRequiringQuotes) != -1
+                                    ? "'" + viewDefinition.name + "'"
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Done.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 13]
```diff
@@ -5782,11 +5782,13 @@ private static void CompleteFormatViewByInferredType(CompletionContext context,
...
-                                    : viewDefinition.name.IndexOfAny(s_charactersRequiringQuotes) != -1
-                                        ? "'" + viewDefinition.name + "'"
-                                        : viewDefinition.name;
+                                string completionText = viewDefinition.name;
+                                var quoteInUse = quote == string.Empty ? "'" : quote;
+                                if (quoteInUse == "'")
+                                {
+                                    completionText = completionText.Replace("'", "''");
+                                }
+                                completionText = quoteInUse + completionText + quoteInUse;
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5791]: This is the same as this right?

```suggestion
                                // If we need quotes, convert to single quotes and escape any existing ones
                                string completionText = quote == string.Empty
                                    ? viewDefinition.name
                                    : "'" + viewDefinition.name.Replace("'", "''") + "'";
```

Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5791]: Or if we need to keep the check for characters in need of escaping:

```suggestion
                                 // If we need quotes, convert to single quotes and escape any existing ones
                                string completionText = (quote == string.Empty && viewDefinition.name.IndexOfAny(s_charactersRequiringQuotes) < 0)
                                    ? viewDefinition.name
                                    : "'" + viewDefinition.name.Replace("'", "''") + "'";
```

**Comment by iSazonov** [comment]
```powershell
        It 'Should complete $processList = Get-Process; $processList | <cmd>' -TestCases (
            @{ cmd = 'Format-Table -View '; expected = "'R A M'", "Priority", "process", "ProcessModule", "ProcessWithUserName", "StartTime" },
```
I still don't understand why this test only crashed on Windows.
I directly add in the test follow:
```powershell
$a="'R A M'", "Priority", "process", "ProcessModule", "ProcessWithUserName", "StartTime"
$a | sort | write-verbose
```
and did get:
```
Priority
process
ProcessModule
ProcessWithUserName
'R A M'
StartTime
```
On all other platform CIs (Linux and MacOS) and on my local Windows system the result was as expected:
```
'R A M'
Priority
process
ProcessModule
ProcessWithUserName
StartTime
```
I don't found why Sort-Object (and C# Array.Sort() - this I tested too) works so weird on Windows CI and add a workaround (sort expected list too).
Does anyone understand what the cause of the problem is?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 13 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: I think the logic here still needs reworking

**Review by rjmholt — COMMENTED** [review]
The tests look good, but that quoting logic I think could be simpler

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 13 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: I use the pattern which we use in other places. I think it makes code more understandable.

**Comment by iSazonov** [comment]
I opened new issue in .Net Runtime to get understanding about the weird sort behavior.
Although the latter tests is more stable and we may not wait for an answer there.

**Comment by iSazonov** [comment]
https://github.com/dotnet/runtime/issues/54472#issuecomment-867108314

Our CI-Windows uses old Windows version without full ICU support. The link above contains a workaround if we want to have a consistency on all Windows 10 systems and use ICU instead of Windows NLS.

**Review by SteveL-MSFT — APPROVED** [review]

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 14]
```diff
@@ -5697,6 +5744,60 @@ private static bool IsInDscContext(ExpressionAst expression)
...
+                                && uniqueNames.Add(viewDefinition.name)
+                                && viewPattern.IsMatch(viewDefinition.name))
+                            {
+                                string completionText = viewDefinition.name;
+                                var quoteInUse = quote == string.Empty && viewDefinition.name.IndexOfAny(s_charactersRequiringQuotes) != -1 ? "'" : quote;
+                                if (quoteInUse == "'")
+                                {
+                                    completionText = completionText.Replace("'", "''");
+                                }
+                                completionText = quoteInUse + completionText + quoteInUse;
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5791]:
```suggestion
                                // If the string is quoted or if it contains characters that need quoting, quote it in single quotes
                                if (quote != string.Empty || viewDefinition.name.IndexOfAny(s_charactersRequiringQuotes) != -1)
                                {
                                    completionText = "'" + completionText.Replace("'", "''") + "'";
                                }
```

**Comment by rjmholt** [comment]
@PoshChan please remind me in 4 hours

**Comment by PoshChan** [comment]
@rjmholt, this is the reminder you requested 4 hours ago
